guides/source/active_model_basics.md

```diff
@@ -44,76 +66,332 @@ class EmailContact
+# casts the string to a date set by the attribute
+irb> person.date_of_birth = "2020-01-01"
+irb> person.date_of_birth
+=> Wed, 01 Jan 2020
+irb> person.date_of_birth.class
+=> Date
+
+# obtains the default value set by the attribute
+irb> person.active
 => true
-irb> email_contact.persisted?
+
+# casts the string to a boolean set by the attribute
+irb> person.active = "0"
+irb> person.active
 => false
 ```
 
-Any class that includes `ActiveModel::API` can be used with `form_with`,
-`render` and any other Action View helper methods, just like Active Record
-objects.
+Some additional methods described below are available when using
+`ActiveModel::Attributes`.
+
+#### Method: `attribute_names`
+
+The `attribute_names` method returns an array of attribute names.
+
+```irb
+irb> Person.attribute_names
+=> ["name", "date_of_birth", "active"]
+```
+
+#### Method: `attributes`
+
+The `attributes` method returns a hash of all the attributes with their names as
+keys and the values of the attributes as values.
+
+```irb
+irb> person = Person.new
+irb> person.name = "John"
+irb> person.date_of_birth = "1998-01-01"
+irb> person.active = false
+
+irb> person.attributes
+=> {"name"=>"John", "date_of_birth"=>Thu, 01 Jan 1998, "active"=>false}
+```
+
+### Attribute Assignment
+
+`ActiveModel::AttributeAssignment` allows you to set an object's attributes by
+passing in a hash of attributes with keys matching the attribute names. This is
+useful when you want to set multiple attributes at once.
+
+Consider the following class:
+
+```ruby
+class Person
+  include ActiveModel::AttributeAssignment
+
+  attr_accessor :name, :date_of_birth, :active
+end
+```
+
+You can set multiple attributes at once using the `assign_attributes` method:
+
+```irb
+irb> person = Person.new
+
+# using the `assign_attributes` method to set multiple attributes at once
+irb> person.assign_attributes(name: "John", date_of_birth: "1998-01-01", active: false)
+
+# check the values of the attributes that were assigned
+irb> person.name
+=> "John"
+irb> person.date_of_birth
+=> Thu, 01 Jan 1998
+irb> person.active
+=> false
+```
+
+If the passed hash responds to the `permitted?` method and the return value of
+this method is `false`, an `ActiveModel::ForbiddenAttributesError` exception is
+raised.
+
+#### Method alias: `attributes=`
+
+The `assign_attributes` method has an alias `attributes=`.
+
+INFO: A method alias is a method that performs the same action as another
+method, but is called something different.
+
+The following example demonstrates the use of the `attributes=` method to set
+multiple attributes at once:
+
+```irb
+irb> person = Person.new
+
+# using the `attributes= method` to set multiple attributes at once
+irb> person.attributes = { name: "John", date_of_birth: "1998-01-01", active: false }
+
+irb> person.name
+=> "John"
+
+# using the `assign_attributes` method to set multiple attributes at once
+irb> person.assign_attributes(name: "Jane")
+
+irb> person.name
+=> "Jane"
```

Comment:
Question: why does `assign_attributes` use () but `attributes=` uses {} to pass the same attributes? Does the `attributes=` alias only accept hashes? And if so, should this be mentioned?

Reply:
Both are method calls, but while `assign_attributes` is generally seen as a "normal" method call which may or may not contain `()`, `attributes=` is commonly considered an attribute/property "setter" and does not use `()`. Any value can be passed to setters, not only Hashes using `{}`. (it just so happens that hashes may or may not require the explicit `{}`, it's a Ruby language thing)

In fact, `assign_attributes` is actually being called with a hash the same way that `attributes=` is, except the `{}` can be omitted in arguments for method calls most of the time, but this would be totally valid: 

```ruby
>> person.assign_attributes({ name: "Jane" })
=> {:name=>"Jane"}
```

And `attributes=` can be called with `()` as well:

```ruby
>> person.attributes = ({ name: "John" })
=> {:name=>"John"}

# in this case, however, Ruby doesn't like omitting `{}`:
>> person.attributes=(name: "John")
(irb):46: syntax error, unexpected label (SyntaxError)
```

Since both `assign_attributes` and `attributes=` are essentially the same, they _expect_ a hash as argument. But that's not a requirement for methods / setters.

Reply:
> Question: why does `assign_attributes` use () but `attributes=` uses {} to pass the same attributes? Does the `attributes=` alias only accept hashes? And if so, should this be mentioned?

That is a really good question!

Reply:
I love this question and I think a lot of folks would not know this offhand. I'd love to add it to the guides but I'm trying to decide whether to add a little INFO block here, or maybe since this can apply in many contexts, I can add a miscellaneous section, any thoughts?  I'm leaning towards a miscellaneous section.